clang < 20.1.0 with cuda incorrect ptx codegen due to compiler bug

Likely related to this: https://github.com/llvm/llvm-project/pull/113216 which was fixed in clang 20.1.0. There are recommended compile flags for working around this but I haven't been able to get them to work yet.

```
ptxas /tmp/TestStdAlgorithmsTeamSwapRanges-9fd836/TestStdAlgorithmsTeamSwapRanges-sm_61.s, line 480426; fatal   : Parsing error near '.': syntax error
ptxas fatal   : Ptx assembly aborted due to errors
clang: error: ptxas command failed with exit code 255 (use -v to see invocation)
Ubuntu clang version 15.0.7
Target: x86_64-pc-linux-gnu
Thread model: posix
InstalledDir: /bin
clang: note: diagnostic msg: 
********************

PLEASE ATTACH THE FOLLOWING FILES TO THE BUG REPORT:
Preprocessed source(s) and associated run script(s) are located at:
clang: note: diagnostic msg: /tmp/TestStdAlgorithmsTeamSwapRanges-2f129e.cu
clang: note: diagnostic msg: /tmp/TestStdAlgorithmsTeamSwapRanges-e6cc03/TestStdAlgorithmsTeamSwapRanges-sm_61.cu
clang: note: diagnostic msg: /tmp/TestStdAlgorithmsTeamSwapRanges-2f129e.sh
clang: note: diagnostic msg: 

********************
gmake[2]: *** [algorithms/unit_tests/CMakeFiles/Kokkos_AlgorithmsUnitTest_StdSet_Team_M.dir/build.make:160: algorithms/unit_tests/CMakeFiles/Kokkos_AlgorithmsUnitTest_StdSet_Team_M.dir/TestStdAlgorithmsTeamSwapRanges.cpp.o] Error 1
```

--- a/algorithms/src/std_algorithms/Kokkos_SwapRanges.hpp
+++ b/algorithms/src/std_algorithms/Kokkos_SwapRanges.hpp
@@ -44,7 +44,7 @@ auto swap_ranges(const ExecutionSpace& ex,
   Impl::static_assert_is_admissible_to_kokkos_std_algorithms(source);
   Impl::static_assert_is_admissible_to_kokkos_std_algorithms(dest);
 
-  assert(source.extent(0) == dest.extent(0));
+  KOKKOS_ASSERT(source.extent(0) == dest.extent(0));
   return Impl::swap_ranges_exespace_impl("Kokkos::swap_ranges_view_api_default",
                                          ex, begin(source), end(source),
                                          begin(dest));
@@ -68,7 +68,7 @@ auto swap_ranges(const std::string& label, const ExecutionSpace& ex,
   Impl::static_assert_is_admissible_to_kokkos_std_algorithms(source);
   Impl::static_assert_is_admissible_to_kokkos_std_algorithms(dest);
 
-  assert(source.extent(0) == dest.extent(0));
+  KOKKOS_ASSERT(source.extent(0) == dest.extent(0));
   return Impl::swap_ranges_exespace_impl(label, ex, begin(source), end(source),
                                          begin(dest));
 }
@@ -98,7 +98,7 @@ KOKKOS_FUNCTION auto swap_ranges(
   Impl::static_assert_is_admissible_to_kokkos_std_algorithms(source);
   Impl::static_assert_is_admissible_to_kokkos_std_algorithms(dest);
 
-  assert(source.extent(0) == dest.extent(0));
+  KOKKOS_ASSERT(source.extent(0) == dest.extent(0));
   return Impl::swap_ranges_team_impl(teamHandle, begin(source), end(source),
                                      begin(dest));
 }

--- a/algorithms/src/std_algorithms/impl/Kokkos_RemoveAllVariants.hpp
+++ b/algorithms/src/std_algorithms/impl/Kokkos_RemoveAllVariants.hpp
@@ -126,7 +126,7 @@ IteratorType remove_if_exespace_impl(const std::string& label,
         func1_type(first, begin(tmp_view), pred), scan_count);
 
     // scan_count should be equal to keep_count
-    assert(scan_count == keep_count);
+    KOKKOS_ASSERT(scan_count == keep_count);
     (void)scan_count;  // to avoid unused complaints
 
     // stage 2, we do parfor to move from tmp to original range

--- a/algorithms/unit_tests/TestStdAlgorithmsTeamIsSortedUntil.cpp
+++ b/algorithms/unit_tests/TestStdAlgorithmsTeamIsSortedUntil.cpp
@@ -158,7 +158,7 @@ void test_A(std::size_t numTeams, std::size_t numCols, int apiId,
        this allows us to exercise that the algorithm returns
        the larest sorted interval starting from 0
     */
-    assert(numCols > 10);
+    KOKKOS_ASSERT(numCols > 10);
     const std::size_t midPoint = numCols / 2;
 
     UnifDist<int> randPoolA(0, midPoint, 3432779);

--- a/algorithms/unit_tests/TestStdAlgorithmsTransformExclusiveScan.cpp
+++ b/algorithms/unit_tests/TestStdAlgorithmsTransformExclusiveScan.cpp
@@ -71,24 +71,24 @@ void fill_view(ViewType dest_view, const std::string& name) {
   }
 
   else if (name == "one-element") {
-    assert(v_h.extent(0) == 1);
+    KOKKOS_ASSERT(v_h.extent(0) == 1);
     v_h(0) = static_cast<value_type>(1);
   }
 
   else if (name == "two-elements-a") {
-    assert(v_h.extent(0) == 2);
+    KOKKOS_ASSERT(v_h.extent(0) == 2);
     v_h(0) = static_cast<value_type>(1);
     v_h(1) = static_cast<value_type>(2);
   }
 
   else if (name == "two-elements-b") {
-    assert(v_h.extent(0) == 2);
+    KOKKOS_ASSERT(v_h.extent(0) == 2);
     v_h(0) = static_cast<value_type>(2);
     v_h(1) = static_cast<value_type>(-1);
   }
 
   else if (name == "small-a") {
-    assert(v_h.extent(0) == 9);
+    KOKKOS_ASSERT(v_h.extent(0) == 9);
     v_h(0) = static_cast<value_type>(3);
     v_h(1) = static_cast<value_type>(1);
     v_h(2) = static_cast<value_type>(4);
@@ -101,7 +101,7 @@ void fill_view(ViewType dest_view, const std::string& name) {
   }
 
   else if (name == "small-b") {
-    assert(v_h.extent(0) >= 6);
+    KOKKOS_ASSERT(v_h.extent(0) >= 6);
     for (std::size_t i = 0; i < ext; ++i) {
       v_h(i) = randObj();
     }

--- a/algorithms/unit_tests/TestStdAlgorithmsTransformInclusiveScan.cpp
+++ b/algorithms/unit_tests/TestStdAlgorithmsTransformInclusiveScan.cpp
@@ -71,24 +71,24 @@ void fill_view(ViewType dest_view, const std::string& name) {
   }
 
   else if (name == "one-element") {
-    assert(v_h.extent(0) == 1);
+    KOKKOS_ASSERT(v_h.extent(0) == 1);
     v_h(0) = static_cast<value_type>(1);
   }
 
   else if (name == "two-elements-a") {
-    assert(v_h.extent(0) == 2);
+    KOKKOS_ASSERT(v_h.extent(0) == 2);
     v_h(0) = static_cast<value_type>(1);
     v_h(1) = static_cast<value_type>(2);
   }
 
   else if (name == "two-elements-b") {
-    assert(v_h.extent(0) == 2);
+    KOKKOS_ASSERT(v_h.extent(0) == 2);
     v_h(0) = static_cast<value_type>(2);
     v_h(1) = static_cast<value_type>(-1);
   }
 
   else if (name == "small-a") {
-    assert(v_h.extent(0) == 9);
+    KOKKOS_ASSERT(v_h.extent(0) == 9);
     v_h(0) = static_cast<value_type>(3);
     v_h(1) = static_cast<value_type>(1);
     v_h(2) = static_cast<value_type>(4);
@@ -101,7 +101,7 @@ void fill_view(ViewType dest_view, const std::string& name) {
   }
 
   else if (name == "small-b") {
-    assert(v_h.extent(0) >= 6);
+    KOKKOS_ASSERT(v_h.extent(0) >= 6);
     for (std::size_t i = 0; i < ext; ++i) {
       v_h(i) = randObj();
     }

--- a/core/src/Kokkos_Tuners.hpp
+++ b/core/src/Kokkos_Tuners.hpp
@@ -338,7 +338,7 @@ class MultidimensionalSparseTuningProblem {
   MultidimensionalSparseTuningProblem(StoredProblemSpace space,
                                       const std::vector<std::string>& names)
       : m_space(std::move(space)), m_variable_names(names) {
-    assert(names.size() == space_dimensionality);
+    KOKKOS_ASSERT(names.size() == space_dimensionality);
     for (unsigned long x = 0; x < names.size(); ++x) {
       VariableInfo info;
       info.type = Kokkos::Tools::Experimental::ValueType::kokkos_value_double;

--- a/core/src/OpenMP/Kokkos_OpenMP_Task.hpp
+++ b/core/src/OpenMP/Kokkos_OpenMP_Task.hpp
@@ -91,7 +91,7 @@ class TaskQueueSpecialization<SimpleTaskScheduler<Kokkos::OpenMP, QueueType>> {
         0,                                    /* team shared buffer */
         0                                     /* thread local buffer */
     );
-    assert(pool_size % team_size == 0);
+    KOKKOS_ASSERT(pool_size % team_size == 0);
 
     auto& queue = scheduler.queue();
 
@@ -257,7 +257,7 @@ class TaskQueueSpecializationConstrained<
         ,
         0 /* thread local buffer */
     );
-    assert(pool_size % team_size == 0);
+    KOKKOS_ASSERT(pool_size % team_size == 0);
 
     auto& queue = scheduler.queue();
     queue.initialize_team_queues(pool_size / team_size);

--- a/core/src/SYCL/Kokkos_SYCL_Instance.cpp
+++ b/core/src/SYCL/Kokkos_SYCL_Instance.cpp
@@ -357,7 +357,7 @@ SYCLInternal::IndirectKernelMem& SYCLInternal::get_indirect_kernel_mem() {
 
 template <sycl::usm::alloc Kind>
 size_t SYCLInternal::USMObjectMem<Kind>::reserve(size_t n) {
-  assert(m_q);
+  KOKKOS_ASSERT(m_q);
 
   if (m_capacity < n) {
     AllocationSpace alloc_space(*m_q);

--- a/core/src/SYCL/Kokkos_SYCL_Instance.hpp
+++ b/core/src/SYCL/Kokkos_SYCL_Instance.hpp
@@ -151,9 +151,10 @@ class SYCLInternal {
     }
 
     void register_event(sycl::event event) {
-      assert(m_last_event
-                 .get_info<sycl::info::event::command_execution_status>() ==
-             sycl::info::event_command_status::complete);
+      KOKKOS_ASSERT(
+          m_last_event
+              .get_info<sycl::info::event::command_execution_status>() ==
+          sycl::info::event_command_status::complete);
       m_last_event = event;
       m_mutex.unlock();
     }

--- a/core/src/impl/Kokkos_Core.cpp
+++ b/core/src/impl/Kokkos_Core.cpp
@@ -265,7 +265,7 @@ int Kokkos::Impl::get_ctest_gpu(int local_rank) {
 
   // Make sure rank is within bounds of resource groups specified by CTest
   auto resource_group_count = std::stoi(ctest_resource_group_count_str);
-  assert(local_rank >= 0);
+  KOKKOS_ASSERT(local_rank >= 0);
   if (local_rank >= resource_group_count) {
     std::ostringstream ss;
     ss << "Error: local rank " << local_rank

--- a/core/src/impl/Kokkos_TaskQueueMultiple.hpp
+++ b/core/src/impl/Kokkos_TaskQueueMultiple.hpp
@@ -209,7 +209,7 @@ class LeagueQueueCollection {
   void initialize_team_queues(int arg_count,
                               memory_pool const& arg_memory_pool) noexcept {
     arg_count = std::min((int)max_num_queues, arg_count);
-    // assert(arg_count <= max_num_queues);
+    // KOKKOS_ASSERT(arg_count <= max_num_queues);
     if (arg_count > m_size) {
       for (int i = m_size; i < arg_count; ++i) {
         new (&m_queues[i - 1].initialized)
@@ -231,9 +231,9 @@ class LeagueQueueCollection {
   team_queue_type& get_team_queue(int iteam) {
     iteam %= max_num_queues;
 #if !defined(__HIP_DEVICE_COMPILE__) && !defined(__CUDA_ARCH__)
-    assert(initialized());
-    assert(iteam < m_size);
-    assert(iteam >= 0);
+    KOKKOS_ASSERT(initialized());
+    KOKKOS_ASSERT(iteam < m_size);
+    KOKKOS_ASSERT(iteam >= 0);
 #endif
     if (iteam == 0)
       return *m_rank_zero_queue;

--- a/core/unit_test/TestTaskScheduler.hpp
+++ b/core/unit_test/TestTaskScheduler.hpp
@@ -281,7 +281,7 @@ struct TestTaskTeam {
                          parscan_check, begin - 1));
 
 #if !defined(__HIP_DEVICE_COMPILE__) && !defined(__CUDA_ARCH__)
-        assert(!future.is_null());
+        KOKKOS_ASSERT(!future.is_null());
 #endif
 
         Kokkos::respawn(this, future);
@@ -485,7 +485,7 @@ struct TestTaskTeamValue {
                                   Kokkos::TaskTeam);
 
 #if !defined(__HIP_DEVICE_COMPILE__) && !defined(__CUDA_ARCH__)
-        assert(!future.is_null());
+        KOKKOS_ASSERT(!future.is_null());
 #endif
 
         sched.respawn(this, future);

--- a/core/unit_test/TestViewCopy_c.hpp
+++ b/core/unit_test/TestViewCopy_c.hpp
@@ -24,7 +24,7 @@ template <class ViewType, class ExecSpace>
 void reset_view(const ExecSpace& space, ViewType& a, int magic) {
   auto policy = Kokkos::RangePolicy<ExecSpace>(space, 0, a.span());
 
-  assert(a.span_is_contiguous());
+  KOKKOS_ASSERT(a.span_is_contiguous());
 
   Kokkos::parallel_for(
       "TestViewCopy::ResetView", policy,
@@ -35,7 +35,7 @@ template <class ViewType, class ExecSpace>
 size_t compute_overall_sum(const ExecSpace& space, ViewType& a) {
   auto policy = Kokkos::RangePolicy<ExecSpace>(space, 0, a.span());
 
-  assert(a.span_is_contiguous());
+  KOKKOS_ASSERT(a.span_is_contiguous());
 
   typename ViewType::value_type sum = 0;
   Kokkos::parallel_reduce(

--- a/core/unit_test/UnitTest_DeviceAndThreads.cpp
+++ b/core/unit_test/UnitTest_DeviceAndThreads.cpp
@@ -34,7 +34,7 @@ int get_num_devices() {
 #else
   num_devices = -1;
 #endif
-  assert(num_devices == Kokkos::num_devices());
+  KOKKOS_ASSERT(num_devices == Kokkos::num_devices());
   return num_devices;
 }
 
@@ -57,7 +57,7 @@ int get_device_id() {
 #else
   device_id   = -1;
 #endif
-  assert(device_id == Kokkos::device_id());
+  KOKKOS_ASSERT(device_id == Kokkos::device_id());
   return device_id;
 }
 
@@ -81,7 +81,7 @@ int get_hwloc_enabled() {
 
 int get_num_threads() {
   int const num_threads = Kokkos::DefaultHostExecutionSpace().concurrency();
-  assert(num_threads == Kokkos::num_threads());
+  KOKKOS_ASSERT(num_threads == Kokkos::num_threads());
   return num_threads;
 }
 

--- a/core/unit_test/tools/TestBuiltinTuners.cpp
+++ b/core/unit_test/tools/TestBuiltinTuners.cpp
@@ -74,11 +74,11 @@ int main(int argc, char* argv[]) {
      * At 0.9 this will be the last element (5)
      */
     auto begin_point = new_team_tuner.get_point(0.0, 0.0, 0.0);
-    assert(std::get<0>(begin_point) == 1);
+    KOKKOS_ASSERT(std::get<0>(begin_point) == 1);
     (void)begin_point;  // to avoid warnings in some compilers
     auto end_point = new_team_tuner.get_point(0.9, 0.0, 0.0);
     (void)end_point;  // to avoid warnings in some compilers
-    assert(std::get<0>(end_point) == 5);
+    KOKKOS_ASSERT(std::get<0>(end_point) == 5);
     for (int x = 0; x < 10000; ++x) {
       auto config = new_team_tuner.begin();
       int option  = std::get<0>(config);

